### ISSUE TITLE
Making regionprops 3D

### ISSUE THREAD

**Issue by emmanuelle** [scikit-image/scikit-image#1489]
`measure.regionprops` currently only works for 2-D images. I would like to extend `regionprops` to work on 3-D arrays as well. Before I start working on a pull request, I'd like to get some feedback on the best way to do it. I was thinking of creating a  `_RegionProperties3D` class that would inherit from `_RegionProperties`. It would also possible to modify some methods of `_RegionProperties` to make them dimension-independent. 

One issue is that some computations are difficult to do in 3D (like Euler number), so some methods would have to return "not implemented" or an error for 3D images.

Any thoughts on porting `regionprops` to 3D images?

**Comment by jni** [comment]
Ooh ooh, pick me! I have thoughts! =P

1. I'm very :-1: on a new class.
2. I think all properties can be extended to 3D with varying amounts of effort.
3. Some are trivial (e.g. area), so they can be dispatched no matter what the input. Others, like moments, are a bit harder. Either the function itself can be made dimension-agnostic, or we can dispatch to different 2D or 3D functions depending on the input. I obviously prefer the former, but it requires more thought.
4. Because of the lazy evaluation of regionprops, we (ie you =P) don't have to implement everything in one PR. You can just add a few functions, and make the others raise a `NotImplementedError` upon receiving 3D input.

**Comment by emmanuelle** [comment]
@jni thanks a lot for your thoughts! Area is in fact not that trivial since it calls the ``moments`` method :-).

How would you handle the case of functions not implemented for 3D? By raising an error in each of these functions, or by adding another decorator? Or something else?

**Comment by jni** [comment]
Huh??? That's a ridiculous implementation! It's just the sum of the image! 8-O

I imagine if you want both the area *and* the moments, it's wasteful not to use the moments, but it's nothing in comparison to computing the moments when you only want the area!

I'm honestly not sure what the best way is to implement. Decorator might not be a bad idea, implement once and add anywhere. `@notimplemented3d`?

**Comment by emmanuelle** [comment]
it is, isn't it? And it's make the computation of the area quite slow as a result (as compared to ``ndimage``, for example). I'll modify ``area``, then.

**Comment by emmanuelle** [comment]
Yeah, about a ``@notimplemented3d``, it would reduce code duplication, but also decrease the code readability. Also, Gaël tells me that decorating functions can break parallel computing if functions cannot be pickled any more. I'm not that familiar with decorators, so we might need another advice. @stefanv @ahojnnes ?

**Comment by stefanv** [comment]
@emmanuelle I wasn't aware of this caveat--can you provide more detail?

**Comment by emmanuelle** [comment]
Gaël has been writing about it in http://gael-varoquaux.info/programming/decoration-in-python-done-right-decorating-and-pickling.html
disclaimer: this blog post dates back to 2009 and the caveat might not exist any more.

**Comment by stefanv** [comment]
That's great, because he also posts a solution :)

**Comment by jni** [comment]
We've used `functools.wraps` in scikit-image before (see e.g. [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/morphology/misc.py#L33), but it's not the only example), so it fits nicely in our codebase. =)

**Comment by soupault** [comment]
Adding reference to https://github.com/scikit-image/scikit-image/issues/1092 .

**Comment by jmetz** [comment]
Just made a really minor change to `centroid` to use the `coords` instead of `moments` (via `local_centroid`) - #2083 - let me know if this is sensible or not... seems like an easy fix to me.
